### PR TITLE
[WIP: Review Needed] Made Plugin Updates Async

### PR DESCRIPTION
**Pull request by betterclever**
This patch removes the Synchronous Plugin Updates API in favour of new Async API using Notifications added in #367 

Tracking Issue: #466

### CODE REVIEW

File: rust/core-lib/src/core.rs [hunk 1]
```diff
@@ -144,8 +144,8 @@ impl WeakXiCore {
     /// Handles the result of an update sent to a plugin.
     ///
     /// All plugins must acknowledge when they are sent a new update, so that
-    /// core can track which revisiions are still 'live', that is can still
-    /// be the base revision for a delta. Once a plugin has acknowleged a new
```
Comment (cmyr) [inline rust/core-lib/src/core.rs]: 👍 this is embarrassing :)

File: rust/plugin-lib/src/lib.rs [hunk 2]
```diff
@@ -109,7 +109,7 @@ pub trait Plugin {
     /// Called when an edit has occured in the remote view. If the plugin wishes
     /// to add its own edit, it may return `Some(edit)`.
     fn update(&mut self, view: &mut View<Self::Cache>, delta: Option<&RopeDelta>,
-              edit_type: String, author: String) -> Option<PluginEdit>;
+              edit_type: String, author: String) -> Option<u64>;
```
Comment (cmyr) [inline rust/plugin-lib/src/lib.rs]: This can now return nothing; the plugin lib can handle acking the update without the plugin needing to know about it.

File: rust/syntect-plugin/src/main.rs [hunk 3]
```diff
@@ -223,10 +222,10 @@ impl<'a> Syntect<'a> {
     /// Checks if a newline has been inserted, and if so inserts whitespace
     /// as necessary.
     fn do_indentation(&mut self, view: &mut MyView, start: usize,
-                      end: usize, text: &str) -> Option<PluginEdit> {
+                      end: usize, text: &str) -> Option<u64> {
```
Comment (cmyr) [inline rust/syntect-plugin/src/main.rs]: ditto, since we don't need to return an edit we don't need to return anything.

File: rust/syntect-plugin/src/main.rs [hunk 4]
```diff
@@ -249,15 +249,24 @@ impl<'a> Syntect<'a> {
...
                 let mut builder = EditBuilder::new(buf_size);
                 builder.replace(interval, indent.into());
+
                 let delta = builder.build();
-                return Some(PluginEdit {
-                    rev: rev,
-                    delta: delta,
+                Some(delta)
+            } else {
+                None
```
Comment (cmyr) [inline rust/syntect-plugin/src/main.rs]: could we not just return in the case that there's no delta? Then `result` doesn't need to be an option, and the following code is a bit cleaner.

File: rust/syntect-plugin/src/main.rs [hunk 5]
```diff
@@ -317,20 +326,19 @@ impl<'a> Plugin for Syntect<'a> {
         self.do_highlighting(view);
     }
 
-    fn config_changed(&mut self, _view: &mut View<Self::Cache>, _changes: &ConfigTable) {
-    }
+    fn config_changed(&mut self, _view: &mut View<Self::Cache>, _changes: &ConfigTable) {}
 
     fn update(&mut self, view: &mut View<Self::Cache>, delta: Option<&RopeDelta>,
-              _edit_type: String, _author: String) -> Option<PluginEdit> {
+              _edit_type: String, _author: String) -> Option<u64> {
```
Comment (cmyr) [inline rust/syntect-plugin/src/main.rs]: ditto not needing to return anything here

File: rust/Cargo.lock [hunk 6]
```diff
@@ -23,7 +23,7 @@ version = "1.0.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
```
Comment (cmyr) [inline rust/Cargo.lock]: I would remove the Cargo.lock update from this PR, we can update it separately later.

File: rust/core-lib/src/event_context.rs [hunk 7]
```diff
@@ -368,14 +368,8 @@ impl<'a> EventContext<'a> {
     pub(crate) fn do_plugin_update(&mut self, update: Result<Value, RpcError>,
```
Comment (cmyr) [inline rust/core-lib/src/event_context.rs]: we can get rid of the TODO above :)

File: rust/core-lib/src/event_context.rs [hunk 8]
```diff
@@ -368,14 +368,8 @@ impl<'a> EventContext<'a> {
...
 
-        match update.map(serde_json::from_value::<UpdateResponse>) {
-            Ok(Ok(UpdateResponse::Edit(edit))) => {
-                let author = edit.author.clone();
-                self.editor.borrow_mut().apply_plugin_edit(edit, Some(undo_group));
-                self.after_edit(&author);
-                self.render_if_needed();
-            }
-            Ok(Ok(UpdateResponse::Ack(_))) => (),
+        match update.map(serde_json::from_value::<u64>) {
```
Comment (cmyr) [inline rust/core-lib/src/event_context.rs]: here we don't actually need to parse the value at all; as long as this is _any_ `Ok(_)` we should consider the message ack'd.

File: rust/plugin-lib/src/view.rs [hunk 9]
```diff
@@ -131,6 +131,15 @@ impl<C: Cache> View<C> {
         self.peer.send_rpc_notification("add_scopes", &params);
     }
 
+    pub fn edit(&self, edit: PluginEdit) {
```
Comment (cmyr) [inline rust/plugin-lib/src/view.rs]: instead of taking a `PluginEdit`, this should take just the arguments determined by the plugin (such as the delta, priority, etc) and then construct the `PluginEdit` in place.

File: rust/syntect-plugin/src/main.rs (hunk 3 above)
Comment (betterclever) [reply rust/syntect-plugin/src/main.rs]: I thought that we needed to return an Int acknowledgement here, but I guess that is not the case. The plugin_lib sends an acknowledgement when plugin receives any event to the plugin. Right?

File: rust/syntect-plugin/src/main.rs (hunk 4 above)
Comment (betterclever) [reply rust/syntect-plugin/src/main.rs]: I could not call the edit function on the view while the view is mutably borrowed for "get_line", thus I used 'if expression' to return the result instead and utilize it later. The same thing I discussed in IRC.

File: rust/syntect-plugin/src/main.rs (hunk 3 above)
Comment (cmyr) [reply rust/syntect-plugin/src/main.rs]: Correct: whenever the plugin receives a new revision from core (via the `update` rpc) the plugin sends an acknowledgement back. The contract here is that the plugin will not subsequently send any edits to core that are based of revisions older than the most recently ack'd.

This is an Int for convenience, but the plugin lib can just fill that in on that side, and the return value has no meaning beyond an ack. It could just as easily be any other type.

Comment (cmyr) [reply rust/syntect-plugin/src/main.rs]: To clarify: because the `View` object actually constructs any edits, we know that once the plugin lib has received the update, and has updated it's internal `rev`, it will no longer be using any previous value. The actual plugin doesn't need to know any of this; it's just bookkeeping. Therefore the plugin-lib can update the plugin without needing a response; the plugin-lib will always respond to the core RPC to acknowledge, independent of any action taken by the particular plugin.

File: rust/core-lib/src/plugins/manager.rs [hunk 10]
```diff
@@ -99,7 +99,8 @@ impl PluginManager {
                 let buffers = self.buffers.clone().to_weak();
                 let mut plugin_ref = plugin.clone();
 
-                plugin.update(&update, move |response| {
+                plugin.update(&update);
+                /*plugin.update(&update, move |response| {
```
Comment (betterclever) [inline rust/core-lib/src/plugins/manager.rs]: @cmyr What was being done in this method earlier? Are similar changes required somewhere else now.

Comment (cmyr) [reply rust/core-lib/src/plugins/manager.rs]: No, this is not currently used and will be removed shortly.

File: rust/core-lib/src/plugins/rpc.rs [hunk 11]
```diff
@@ -75,6 +75,10 @@ pub struct PluginUpdate {
     /// The total number of lines in the document after applying this delta.
     pub new_line_count: usize,
     pub rev: u64,
+    /// The undo_group associated with this update. Each PluginEdit sent by plugin is associated
```
Comment (cmyr) [inline rust/core-lib/src/plugins/rpc.rs]: I would reword the second sentence a little, to say:

"The plugin may pass this value back to core when making an edit, to associate the plugin's edit with this undo group." (then continue with your sentence that starts, "Core uses ...")

File: rust/plugin-lib/src/lib.rs [hunk 12]
```diff
@@ -107,9 +107,9 @@ pub trait Plugin {
 
     //TODO: async edits only; this is here for feature parity during initial hacking
     /// Called when an edit has occured in the remote view. If the plugin wishes
-    /// to add its own edit, it may return `Some(edit)`.
+    /// to add its own edit, it must do so using asynchronously via edit notification.
```
Comment (cmyr) [inline rust/plugin-lib/src/lib.rs]: Good stuff, it's important to keep docs up to date. I would say "via **the** edit notification".

File: rust/plugin-lib/src/lib.rs [hunk 13]
```diff
@@ -107,9 +107,9 @@ pub trait Plugin {
 
     //TODO: async edits only; this is here for feature parity during initial hacking
```
Comment (cmyr) [inline rust/plugin-lib/src/lib.rs]: we can delete this TODO

File: rust/plugin-lib/src/view.rs [hunk 14]
```diff
@@ -131,7 +131,10 @@ impl<C: Cache> View<C> {
         self.peer.send_rpc_notification("add_scopes", &params);
     }
 
-    pub fn edit(&self, edit: PluginEdit) {
+    pub fn edit(&self, delta: RopeDelta, priority: u64, after_cursor: bool,
+                undo_group: Option<usize>, author: String) {
```
Comment (cmyr) [inline rust/plugin-lib/src/view.rs]: This is just about right, but one little change: instead of `undo_group: Option<usize>` I'd have an argument `new_undo_group: bool`; then we just keep track of the undo group internally, like we do with `rev`, and send it to core whenever `new_undo_group` is `false`.

File: rust/syntect-plugin/src/main.rs [hunk 15]
```diff
@@ -222,10 +222,10 @@ impl<'a> Syntect<'a> {
     /// Checks if a newline has been inserted, and if so inserts whitespace
     /// as necessary.
     fn do_indentation(&mut self, view: &mut MyView, start: usize,
-                      end: usize, text: &str) -> Option<u64> {
+                      end: usize, text: &str, undo_group: Option<usize>) {
```
Comment (cmyr) [inline rust/syntect-plugin/src/main.rs]: related to my previous comment: we know in this case that we always want an undo group, so we don't need to pass this through.

File: rust/syntect-plugin/src/main.rs [hunk 16]
```diff
@@ -329,19 +320,18 @@ impl<'a> Plugin for Syntect<'a> {
     fn config_changed(&mut self, _view: &mut View<Self::Cache>, _changes: &ConfigTable) {}
 
     fn update(&mut self, view: &mut View<Self::Cache>, delta: Option<&RopeDelta>,
-              _edit_type: String, _author: String) -> Option<u64> {
+              undo_group: Option<usize>, _edit_type: String, _author: String) {
```
Comment (cmyr) [inline rust/syntect-plugin/src/main.rs]: ditto earlier, all of syntect's edits will be part of the active undo group so we don't need to pass the argument around.

**Comment by betterclever** [comment]
@cmyr Did the suggested changes. 😄

File: rust/syntect-plugin/src/main.rs [hunk 17]
```diff
@@ -249,17 +247,18 @@ impl<'a> Syntect<'a> {
...
-                    author: "syntect".to_owned(),
-                })
+                Some(delta)
+            } else {
+                None
+            };
+
+            if let Some(delta) = result {
+                view.edit(delta, INDENTATION_PRIORITY, false,
+                          false, String::from("syntect".to_owned()));
```
Comment (cmyr) [inline rust/syntect-plugin/src/main.rs]: Didn't notice this earlier, but this is redundant; you can just do `"syntect".into()` or `"syntect".to_string()` _or_ `String::from("syntect")`. Alternatively you could let `view.edit(..)` take a `&str`, and then you can do the conversion internally. Alternatively there's a pattern we use a lot where the function takes a generic `S: AsRef<str>`, see [this nice explanation](https://deterministic.space/elegant-apis-in-rust.html#use-conversion-traits).

File: rust/plugin-lib/src/lib.rs [hunk 18]
```diff
@@ -129,7 +128,7 @@ pub trait Plugin {
                       view: &mut View<Self::Cache>,
                       changes: &ConfigTable);
 
-    /// Called when the runloop is idle, if the plugin has prevoiusly
+    /// Called when the runloop is idle, if the plugin has preciously
```
Comment (raphlinus) [inline rust/plugin-lib/src/lib.rs]: This should be "previously".

**Comment by betterclever** [comment]
Fixed it. Also resolved merge conflicts.

**Comment by cmyr** [comment]
Great, thanks!
